### DISCUSSION TITLE
how can I turn off password encryption in keycloak??

### DISCUSSION THREAD

**Discussion by yeongseoPark** [keycloak/keycloak#28426] [Q&A] [answered]
i wanna use keycloak as auth server. But my client already has encryption logic in frontend. So that requirement he gave me is to just store given password(without encrpytion) when signing up, and compare it with given password when login. I managed to eliminate encryption function after sign up(by making custom `PasswordHashProvider`), but somehow, after first login, password inside database turns into encrypted form.. 
how can I change this logic that encrypts just-string password to encrypted one after login??

**Comment by codespearhead** [comment, marked as answer]
As the [Open Source Foundation for Application Security (OWASP) puts it](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html):

> Hashing and encryption can keep sensitive data safe, but in almost all circumstances, passwords should be hashed, NOT encrypted.

Encrypted passwords can be unencrypted with the correct cipher, whereas hashed passwords can't. You don't want identity providers to know your password, do you? The latter only allows you to check whether some string matches the stored hash after being fed into the same hash function, which is all you need for authentication purposes, as [Auth0 explains](https://auth0.com/blog/hashing-passwords-one-way-road-to-security/):

> In contrast to encryption, hashing is a one-way mechanism. The data that is hashed cannot be practically "unhashed".
> [...]
> [A hash function is a] deterministic function [...] that given the same input always produces the same output. This is vital for authentication since we need to have the guarantee that a given password will always produce the same hash; otherwise, it would be impossible to consistently verify user credentials with this technique.

Not hashing passwords in the server-end renders security useless because it doesn't matter what the password intercepted in transit looks like, be it in plain text, hashed or encrypted, given that it'll be enough for the malicious agent to authenticate in the system.

Keep in mind that the client-side can easily be tampered with since it's out of your control, and that's why you shouldn't do such a thing there. Encryption of data in transit is already handled by TLS/SSL over HTTPS.

By default, Keycloak hashes passwords and only works over HTTPS in production mode.
